PgBouncer Unable to connect to postgres master

- **Which image of the operator are you using?** v1.6.3
- **Where do you run it - cloud or metal? Kubernetes or OpenShift?** AWS K8s(With Istio side car enabled)
- **Are you running Postgres Operator in production?** no
- **Type of issue?** question

I am trying to install the operator with the connection pooler enabled, Initially I had issues with Postgres pods also, the secondary node was unable to talk to the master, that issue is fixed with the help of this ticket #https://github.com/zalando/postgres-operator/issues/772. 
Now I am having issues with pgbouncer which failing to connect with Postgres cluster service.
PgBoncer log

```
exec /bin/pgbouncer /etc/pgbouncer/pgbouncer.ini
2021-06-01 15:00:25.696 UTC [1] LOG kernel file descriptor limit: 65536 (hard: 65536); max_client_conn: 10000, max expected fd use: 10012
2021-06-01 15:00:25.696 UTC [1] LOG listening on 0.0.0.0:5432
2021-06-01 15:00:25.696 UTC [1] LOG listening on [::]:5432
2021-06-01 15:00:25.696 UTC [1] LOG listening on unix:/tmp/.s.PGSQL.5432
2021-06-01 15:00:25.697 UTC [1] LOG process up: PgBouncer 1.12.0, libevent 2.1.11-stable (epoll), adns: c-ares 1.16.1, tls: OpenSSL 1.1.1g  21 Apr 2020
2021-06-01 15:01:25.698 UTC [1] LOG stats: 0 xacts/s, 0 queries/s, in 0 B/s, out 0 B/s, xact 0 us, query 0 us, wait 0 us
2021-06-01 15:02:11.714 UTC [1] LOG C-0x56242d327a70: (nodb)/(nouser)@127.0.0.1:48248 registered new auto-database: db=testdb
2021-06-01 15:02:11.834 UTC [1] WARNING lookup failed: my-test-pg-cluster: result=0
2021-06-01 15:02:25.702 UTC [1] LOG stats: 0 xacts/s, 0 queries/s, in 0 B/s, out 0 B/s, xact 0 us, query 0 us, wait 0 us
2021-06-01 15:02:27.085 UTC [1] WARNING lookup failed: my-test-pg-cluster: result=0
2021-06-01 15:02:31.418 UTC [1] WARNING C-0x56242d327e90: testdb/(nouser)@127.0.0.1:49168 pooler error: pgbouncer cannot connect to server
2021-06-01 15:02:31.422 UTC [1] WARNING C-0x56242d327c80: testdb/(nouser)@127.0.0.1:49166 pooler error: pgbouncer cannot connect to server
2021-06-01 15:02:33.500 UTC [1] WARNING C-0x56242d327c80: testdb/(nouser)@127.0.0.1:49298 pooler error: pgbouncer cannot connect to server
2021-06-01 15:02:33.502 UTC [1] WARNING C-0x56242d327e90: testdb/(nouser)@127.0.0.1:49300 pooler error: pgbouncer cannot connect to server
2021-06-01 15:02:34.531 UTC [1] WARNING C-0x56242d327e90: testdb/(nouser)@127.0.0.1:49334 pooler error: pgbouncer cannot connect to server
2021-06-01 15:02:36.595 UTC [1] WARNING C-0x56242d327e90: testdb/(nouser)@127.0.0.1:49420 pooler error: pgbouncer cannot connect to server
2021-06-01 15:02:36.598 UTC [1] WARNING C-0x56242d327c80: testdb/(nouser)@127.0.0.1:49422 pooler error: pgbouncer cannot connect to server
2021-06-01 15:02:38.641 UTC [1] WARNING C-0x56242d327c80: testdb/(nouser)@127.0.0.1:49520 pooler error: pgbouncer cannot connect to server
2021-06-01 15:02:38.657 UTC [1] WARNING C-0x56242d327e90: testdb/(nouser)@127.0.0.1:49522 pooler error: pgbouncer cannot connect to server
2021-06-01 15:02:40.683 UTC [1] WARNING C-0x56242d327e90: testdb/(nouser)@127.0.0.1:49594 pooler error: pgbouncer cannot connect to server
2021-06-01 15:02:40.698 UTC [1] WARNING C-0x56242d327c80: testdb/(nouser)@127.0.0.1:49596 pooler error: pgbouncer cannot connect to server
2021-06-01 15:02:42.209 UTC [1] WARNING lookup failed: my-test-pg-cluster: result=0

```
I have created a service entry like this

```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: my-test-cluster-service-entry
spec:
  hosts:
    - my-test-pg-cluster-0.dev.svc.cluster.local
    - my-test-pg-cluster-1.dev.svc.cluster.local
  location: MESH_INTERNAL
  ports:
  - number: 5432
    name: postgres
    protocol: TCP
  - number: 8008
    name: tcp-patroni
    protocol: TCP
  resolution: NONE
```

Is there anything I need to add? Any help would be appreciated! 
Thanks in Advance.

Thanks @trezorg, will check it out.
@kien-truong I'm infact seeing some errors in the core DNS logs:
```
[ERROR] plugin/errors: 2 8587284445994789026.1238791110190553003.cluster.local. HINFO: dial tcp x.x.x.x: connect: connection refused
[ERROR] plugin/errors: 2 8587284445994789026.1238791110190553003.cluster.local. HINFO: dial tcp x.x.x.x: connect: connection refused
[ERROR] plugin/errors: 2 3971420788730198820.2231477973618440833.ip6.arpa. HINFO: dial tcp x.x.x.x: connect: connection refused
[ERROR] plugin/errors: 2 3278615996869201988.8959794356365005256.in-addr.arpa. HINFO: dial tcp x.x.x.x: connect: connection refused
[ERROR] plugin/errors: 2 3278615996869201988.8959794356365005256.in-addr.arpa. HINFO: dial tcp x.x.x.x: connect: connection refused
[ERROR] plugin/errors: 2 3971420788730198820.2231477973618440833.ip6.arpa. HINFO: dial tcp x.x.x.x: connect: connection refused
```

Wondering why the DNS resolution only for PG Bouncer to PSQL DB fails.
Communication from our apps to direct PSQL DB via DNS works fine. Other inter-app communication via DNS is also working just fine.

```
[ERROR] plugin/errors: 2 8587284445994789026.1238791110190553003.cluster.local. HINFO: dial tcp x.x.x.x: connect: connection refused
```
This error indicates that `Coredns` is configured to forward DNS queries to `x.x.x.x` , which might not be a valid DNS server, since it's refusing connection. Try changing the `x.x.x.x` to a different upstream DNS server.

`pgbouncer` DNS resolver is problematic because it performs queries in parallel and any `SERVFAIL` responses will immediately trigger failure mode, like this:
```
021-10-29 10:28:02.098 UTC [1] WARNING lookup failed: team-postgresql: result=0
```
This is different than `nslookup`, which performs DNS lookup sequentially until it finds a valid record.

--- a/charts/postgres-operator/crds/operatorconfigurations.yaml
+++ b/charts/postgres-operator/crds/operatorconfigurations.yaml
@@ -597,7 +597,7 @@ spec:
                     default: "pooler"
                   connection_pooler_image:
                     type: string
-                    default: "registry.opensource.zalan.do/acid/pgbouncer:master-22"
+                    default: "registry.opensource.zalan.do/acid/pgbouncer:master-24"
                   connection_pooler_max_db_connections:
                     type: integer
                     default: 60

--- a/charts/postgres-operator/values.yaml
+++ b/charts/postgres-operator/values.yaml
@@ -392,7 +392,7 @@ configConnectionPooler:
   # db user for pooler to use
   connection_pooler_user: "pooler"
   # docker image
-  connection_pooler_image: "registry.opensource.zalan.do/acid/pgbouncer:master-22"
+  connection_pooler_image: "registry.opensource.zalan.do/acid/pgbouncer:master-24"
   # max db connections the pooler should hold
   connection_pooler_max_db_connections: 60
   # default pooling mode

--- a/manifests/configmap.yaml
+++ b/manifests/configmap.yaml
@@ -17,7 +17,7 @@ data:
   # connection_pooler_default_cpu_request: "500m"
   # connection_pooler_default_memory_limit: 100Mi
   # connection_pooler_default_memory_request: 100Mi
-  connection_pooler_image: "registry.opensource.zalan.do/acid/pgbouncer:master-22"
+  connection_pooler_image: "registry.opensource.zalan.do/acid/pgbouncer:master-24"
   # connection_pooler_max_db_connections: 60
   # connection_pooler_mode: "transaction"
   # connection_pooler_number_of_instances: 2

--- a/manifests/minimal-fake-pooler-deployment.yaml
+++ b/manifests/minimal-fake-pooler-deployment.yaml
@@ -23,7 +23,7 @@ spec:
       serviceAccountName: postgres-operator
       containers:
       - name: postgres-operator
-        image: registry.opensource.zalan.do/acid/pgbouncer:master-22
+        image: registry.opensource.zalan.do/acid/pgbouncer:master-24
         imagePullPolicy: IfNotPresent
         resources:
           requests:

--- a/manifests/operatorconfiguration.crd.yaml
+++ b/manifests/operatorconfiguration.crd.yaml
@@ -595,7 +595,7 @@ spec:
                     default: "pooler"
                   connection_pooler_image:
                     type: string
-                    default: "registry.opensource.zalan.do/acid/pgbouncer:master-22"
+                    default: "registry.opensource.zalan.do/acid/pgbouncer:master-24"
                   connection_pooler_max_db_connections:
                     type: integer
                     default: 60

--- a/manifests/postgresql-operator-default-configuration.yaml
+++ b/manifests/postgresql-operator-default-configuration.yaml
@@ -190,7 +190,7 @@ configuration:
     connection_pooler_default_cpu_request: "500m"
     connection_pooler_default_memory_limit: 100Mi
     connection_pooler_default_memory_request: 100Mi
-    connection_pooler_image: "registry.opensource.zalan.do/acid/pgbouncer:master-22"
+    connection_pooler_image: "registry.opensource.zalan.do/acid/pgbouncer:master-24"
     # connection_pooler_max_db_connections: 60
     connection_pooler_mode: "transaction"
     connection_pooler_number_of_instances: 2

--- a/pkg/cluster/k8sres.go
+++ b/pkg/cluster/k8sres.go
@@ -102,8 +102,9 @@ func (c *Cluster) serviceAddress(role PostgresRole) string {
 		return service.ObjectMeta.Name
 	}
 
-	c.logger.Warningf("No service for role %s", role)
-	return ""
+	defaultAddress := c.serviceName(role)
+	c.logger.Warningf("No service for role %s - defaulting to %s", role, defaultAddress)
+	return defaultAddress
 }
 
 func (c *Cluster) servicePort(role PostgresRole) int32 {

